csrc/mmdeploy/net/ipu/ipu_net.cpp

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) OpenMMLab. All rights reserved.
+
+#include "ipu_net.h"
+
+#include <sys/time.h>
+#include <unistd.h>
+
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include "mmdeploy/core/logger.h"
+#include "mmdeploy/core/model.h"
+#include "mmdeploy/core/utils/formatter.h"
+
+namespace mmdeploy::framework {
+
+mmdeploy::DataType IPUNet::ipu_type_convert(const popef::DataType& ipu_type) {
+  mmdeploy::DataType mtype;
+  if (ipu_type == popef::DataType::F32) {
+    mtype = mmdeploy::DataType::kFLOAT;
+  } else if (ipu_type == popef::DataType::F16) {
+    mtype = mmdeploy::DataType::kHALF;
+  } else if (ipu_type == popef::DataType::F8) {
+    mtype = mmdeploy::DataType::kINT8;
+  } else {
+    throw std::invalid_argument(
+        "invalid data type for IPU backend, current legit is : fp32, fp16, fp8");
+  }
+  return mtype;
+}
+
+void IPUNet::copy_output(const model_runtime::TensorMemory& from, Tensor& to) {
+  if (from.data_size_bytes != to.byte_size()) {
+    MMDEPLOY_ERROR("output tensor size not match from size {} to size {}", from.data_size_bytes,
+                   to.byte_size());
+    return;
+  }
+  int size = from.data_size_bytes;
+
+  MMDEPLOY_DEBUG("copy output total byte_size {}", size);
+
+  char* from_ptr = static_cast<char*>(from.data.get());
+
+  char* pto = to.data<char>();
+  for (int i = 0; i < size; i++) {
+    pto[i] = *(from_ptr + i);
+  }
+}
+
+void IPUNet::copy_input(const Tensor& from, model_runtime::TensorMemory& to) {
+  if (from.byte_size() != to.data_size_bytes) {
+    MMDEPLOY_ERROR("input tensor size not match  from size {} to size {} ", from.byte_size(),
+                   to.data_size_bytes);
+    return;
+  }
+  int size = from.byte_size();
+
+  MMDEPLOY_DEBUG("copy input total byte_size {}", size);
+
+  char* to_ptr = static_cast<char*>(to.data.get());
+
+  const char* pfrom = from.data<char>();
+  for (int i = 0; i < size; i++) {
+    *(to_ptr + i) = pfrom[i];
+  }
+}
+
+IPUNet::~IPUNet() {}
+
+Result<void> IPUNet::Init(const Value& args) {
+  auto& context = args["context"];
+  auto name = args["name"].get<std::string>();
+  auto model = context["model"].get<Model>();
+  OUTCOME_TRY(auto config, model.GetModelConfig(name));
+
+  batch_per_step = config.batches_per_step;
```

Comment:
model_meta_info_t don't have batches_per_step filed.

Reply:
When convert the model, you can generate a file named `ipu_params.json`.
The model dir looks like:
```
.
├── ipu_params.json
├── deploy.json
├── detail.json
├── end2end
│   ├── end2end.onnx.optimized.onnx
│   └── executable.popef
├── end2end.onnx
└── pipeline.json
```
 In `ipu_params.json` file you can add params like
```json
{
    "batches_per_step": 1
}
```

And in [csrc/mmdeploy/net/ipu/ipu_net.cpp](https://github.com/open-mmlab/mmdeploy/pull/1570/files#diff-b50ab9a17c2681ad798f351b12cbffd4e8b3e5d88a59f1e7ad80b6e02460e235), you can add a param struct like `model_meta_info_t`
```cpp
struct ipu_params_t {
  int batches_per_step;
  MMDEPLOY_ARCHIVE_MEMBERS(batches_per_step);
};
```

To read the `ipu_params.json`, you can refer to [this](https://github.com/gqingraphcore/mmdeploy-ipu/blob/ipu/csrc/mmdeploy/model/directory_model_impl.cpp#L45-L46)
```cpp
#include "mmdeploy/archive/json_archive.h"

OUTCOME_TRY(auto param_json, model.ReadFile("ipu_params.json"));
ipu_params_t param;
from_json(nlohmann::json::parse(param_json), param);

```

Reply:
revised according to suggestion, but how to test the cpp interface when it requires files dumped from python-converter?